cfme/infrastructure/host.py

```diff
@@ -173,70 +81,129 @@ def __init__(self, **kwargs):
                credentials are invalid, an error will be raised.
         """
-        navigate_to(self, 'Add')
-        fill(properties_form, self._form_mapping(True, **self.__dict__))
-        fill(credential_form, self.credentials, validate=validate_credentials)
-        fill(credential_form, self.ipmi_credentials, validate=validate_credentials)
-        self._submit(cancel, host_add_btn)
+        view = navigate_to(self, "Add")
+        view.fill({
+            "name": self.name,
+            "hostname": self.hostname or self.ip_address,
+            "host_platform": self.host_platform,
+            "custom_ident": self.custom_ident,
+            "ipmi_address": self.ipmi_address,
+            "mac_address": self.mac_address
+        })
+        if self.credentials is not None:
+            view.endpoints.default.fill({
+                "username": self.credentials.principal,
+                "password": self.credentials.secret,
+                "confirm_password": self.credentials.verify_secret,
+            })
+            if validate_credentials:
+                view.endpoints.default.validate_button.click()
+        if self.ipmi_credentials is not None:
+            view.endpoints.fill({
+                "username": self.ipmi_credentials.principal,
+                "password": self.ipmi_credentials.secret,
+                "confirm_password": self.ipmi_credentials.verify_secret,
+            })
+            if validate_credentials:
+                view.endpoints.ipmi.validate_button.click()
+        if not cancel:
+            view.add_button.click()
+            flash_message = 'Host / Node " {}" was added'.format(self.name)
+        else:
+            view.cancel_button.click()
+            flash_message = "Add of new Host / Node was cancelled by the user"
+        view = self.create_view(HostsView)
+        assert view.is_displayed
+        view.flash.assert_success_message(flash_message)
 
     def update(self, updates, cancel=False, validate_credentials=False):
-        """
-        Updates a host in the UI.  Better to use utils.update.update context
-        manager than call this directly.
+        """Updates a host in the UI. Better to use utils.update.update context manager than call
+        this directly.
 
         Args:
            updates (dict): fields that are changing.
-           cancel (boolean): whether to cancel out of the update.
+           cancel (bool): whether to cancel out of the update.
         """
 
-        navigate_to(self, 'Edit')
-        change_stored_password()
-        fill(credential_form, updates.get('credentials'), validate=validate_credentials)
-
-        logger.debug("Trying to save update for host with id: " + str(self.get_db_id))
-        self._submit(cancel, self.forced_saved)
+        view = navigate_to(self, "Edit")
+        changed = view.fill(updates)
+        credentials = updates.get("credentials")
```

Comment:
you need to pop both credentials before changed. Otherwise, you will get error or warning. I encountered this issue in provider.update